flink-table/flink-sql-gateway/src/test/java/org/apache/flink/table/gateway/service/SqlGatewayServiceITCase.java

```diff
@@ -0,0 +1,316 @@
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+
+import static org.apache.flink.core.testutils.FlinkAssertions.anyCauseMatches;
+import static org.apache.flink.types.RowKind.DELETE;
+import static org.apache.flink.types.RowKind.INSERT;
+import static org.apache.flink.types.RowKind.UPDATE_AFTER;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/** ITCase for {@link SqlGatewayServiceImpl}. */
+public class SqlGatewayServiceITCase extends AbstractTestBase {
+
+    @ClassRule
+    public static final SqlGatewayServiceTestResource SQL_GATEWAY_SERVICE_RESOURCE =
+            new SqlGatewayServiceTestResource();
+
+    private static SessionManager sessionManager;
+    private static SqlGatewayService service;
+
+    private final SessionEnvironment defaultSessionEnvironment =
+            SessionEnvironment.newBuilder()
+                    .setSessionEndpointVersion(MockedEndpointVersion.V1)
+                    .build();
+
+    @BeforeClass
+    public static void setup() {
+        sessionManager = SQL_GATEWAY_SERVICE_RESOURCE.getSessionManager();
+        service = SQL_GATEWAY_SERVICE_RESOURCE.getService();
+    }
+
+    @Test
+    public void testOpenSessionWithConfig() {
+        Map<String, String> options = new HashMap<>();
+        options.put("key1", "val1");
+        options.put("key2", "val2");
+        SessionEnvironment environment =
+                SessionEnvironment.newBuilder()
+                        .setSessionEndpointVersion(MockedEndpointVersion.V1)
+                        .addSessionConfig(options)
+                        .build();
+
+        SessionHandle sessionHandle = service.openSession(environment);
+        Map<String, String> actualConfig = service.getSessionConfig(sessionHandle);
+
+        for (String key : options.keySet()) {
+            assertEquals(options.get(key), actualConfig.get(key));
+        }
+    }
+
+    @Test
+    public void testGetOperationFinishedAndFetchResults() throws Exception {
+        SessionHandle sessionHandle = service.openSession(defaultSessionEnvironment);
+
+        CountDownLatch startRunningLatch = new CountDownLatch(1);
+        CountDownLatch endRunningLatch = new CountDownLatch(1);
+
+        OperationHandle operationHandle =
+                submitDefaultOperation(
+                        sessionHandle,
+                        () -> {
+                            startRunningLatch.countDown();
+                            endRunningLatch.await();
+                        });
+
+        startRunningLatch.await();
+        assertEquals(
+                new OperationInfo(OperationStatus.RUNNING, OperationType.UNKNOWN, true),
+                service.getOperationInfo(sessionHandle, operationHandle));
+
+        endRunningLatch.countDown();
+        Thread.sleep(1000);
```

Comment:
Maybe we should use iterations here. Sleep is still not safe when CPU load is high.